templates/decays.cpp.in

```diff
@@ -0,0 +1,301 @@
+   constexpr double ps {1./(8.*Pi)};
+   constexpr double ps_symmetry {1./2.};
+   const double color_gen_sqr = squared_color_generator<H, A, A>();
+   const double flux = 0.5/mH;
+   const auto res = flux * color_gen_sqr * ps * ps_symmetry * amp.square();
+
+   if (res < 0) {
+      throw std::runtime_error("Error in H->AA. Partial width < 0");
+   } else {
+      return res;
+   }
+}
+
+template <>
+double CLASSNAME::get_partial_width<H, A, Z>(
+      const context_base& context,
+      const typename field_indices<H>::type& in_idx,
+      const typename field_indices<A>::type& out1_idx,
+      const typename field_indices<Z>::type& out2_idx) const
+{
+   const double mH = context.physical_mass<H>(in_idx);
+   const double mA = context.physical_mass<A>(out1_idx);
+   const double mZ = context.physical_mass<Z>(out2_idx);
+
+   const double ps {1./(8.*Pi) * sqrt(KallenLambda(1., Sqr(mA/mH), Sqr(mZ/mH)))};
+   constexpr double ps_symmetry {1.};
+   const double color_gen_sqr = squared_color_generator<H, A, Z>();
+   const double flux = 0.5/mH;
+
+   const auto amp = calculate_amplitude<H, A, Z>(context, in_idx, out1_idx, out2_idx);
+
+   const auto res = flux * color_gen_sqr * ps * ps_symmetry * amp.square();
+
+   if (res < 0) {
+      throw std::runtime_error("Error in H->AZ. Partial width < 0");
+   } else {
+      return res;
+   }
+}
+
+template <>
+double CLASSNAME::get_partial_width<H, G, G>(
+      const context_base& context,
+      const typename field_indices<H>::type& in_idx,
+      const typename field_indices<G>::type& out1_idx,
+      const typename field_indices<G>::type& out2_idx) const
+{
+   const auto amp = calculate_amplitude<H, G, G>(context, in_idx, out1_idx, out2_idx);
+   const double mH = context.physical_mass<H>(in_idx);
+   constexpr double ps {1./(8.*Pi)};
+   constexpr double ps_symmetry {1./2.};
+   const double color_fact = squared_color_generator<H, G, G>();
+   const double flux = 0.5/mH;
+
+   double result = flux * color_fact * ps * ps_symmetry * amp.square();
+
+   // higher order corrections to CP-even Higgs
+   if (!info::is_CP_violating_Higgs_sector) {
+      // higher order QCD corrections
+
+      // number of active light flavours
+      unsigned int Nf;
+      const double mtpole {qedqcd.displayPoleMt()};
+      if (mH > 5 && mH < mtpole) {
+         Nf = 5;
+      } else if (mH > mtpole) {
+         Nf = 6;
+      } else {
+         throw;
+      }
+
+      // NLO, NNLO - eq. 11 of hep-ph/9705240
+      const double deltaNLO {95./4. - 7./6.*Nf};
+
+      const double deltaNNLO {
+         149533./288. - 363./8.*zeta2 - 495./8.*zeta3 + 19./8.*std::log(Sqr(mH/mtpole))
+               + Nf*(-4157./72 + 11./2.*zeta2 + 5./4.*zeta3 + 2./3.*std::log(Sqr(mH/mtpole)))
+               + Nf*Nf*(127./108. - zeta2/6.)
+      };
+      // eq. 4.20 from Adam's thesis
+      const double deltaNNNLO {467.683620788 + 122.440972222*std::log(Sqr(mH/mtpole)) + 10.9409722222*Sqr(std::log(Sqr(mH/mtpole)))};
```

Comment:
Please calculate the log only once.

Reply:
Done in 3a46e0bd513a5cfbc9c6c916fa059c9d733b1108.